python_modules/dagster/dagster/core/asset_defs/decorators.py

```diff
@@ -272,99 +259,67 @@ def multi_asset(
         return AssetsDefinition(
-            input_names_by_asset_key={
-                cast(AssetKey, in_def.asset_key): input_name
-                for input_name, in_def in asset_ins.items()
-            },
-            output_names_by_asset_key={
-                cast(AssetKey, out_def.asset_key): output_name for output_name, out_def in asset_outs.items()  # type: ignore
+            asset_keys_by_input_name={
+                input_name: asset_key for asset_key, (input_name, _) in asset_ins.items()
             },
+            asset_keys_by_output_name=asset_keys_by_output_name,
             op=op,
+            asset_deps={asset_keys_by_output_name[name]: asset_deps[name] for name in asset_deps},
         )
 
     return inner
 
 
-def build_asset_outs(
-    op_name: str,
-    outs: Mapping[str, Out],
-    ins: Mapping[str, In],
-    internal_asset_deps: Mapping[str, Set[AssetKey]],
-) -> Dict[str, Out]:
-
-    # if an AssetKey is not supplied, create one based off of the out's name
-    asset_keys_by_out_name = {
-        out_name: out.asset_key if isinstance(out.asset_key, AssetKey) else AssetKey([out_name])
-        for out_name, out in outs.items()
-    }
-
-    # update asset_key if necessary, add metadata indicating inter asset deps
-    outs = {
-        out_name: out._replace(
-            asset_key=asset_keys_by_out_name[out_name],
-            metadata=dict(
-                **(out.metadata or {}),
-                **(
-                    {ASSET_DEPENDENCY_METADATA_KEY: internal_asset_deps[out_name]}
-                    if out_name in internal_asset_deps
-                    else {}
-                ),
-            ),
-        )
-        for out_name, out in outs.items()
-    }
-
-    # validate that the internal_asset_deps make sense
-    valid_asset_deps = set(in_def.asset_key for in_def in ins.values())
-    valid_asset_deps.update(asset_keys_by_out_name.values())
-    for out_name, asset_keys in internal_asset_deps.items():
-        check.invariant(
-            out_name in outs,
-            f"Invalid out key '{out_name}' supplied to `internal_asset_deps` argument for multi-asset "
-            f"{op_name}. Must be one of the outs for this multi-asset {list(outs.keys())}.",
-        )
-        invalid_asset_deps = asset_keys.difference(valid_asset_deps)
-        check.invariant(
-            not invalid_asset_deps,
-            f"Invalid asset dependencies: {invalid_asset_deps} specified in `internal_asset_deps` "
-            f"argument for multi-asset '{op_name}' on key '{out_name}'. Each specified asset key "
-            "must be associated with an input to the asset or produced by this asset. Valid "
-            f"keys: {valid_asset_deps}",
-        )
-
-    return outs
-
-
 def build_asset_ins(
     fn: Callable,
     asset_namespace: Optional[Sequence[str]],
     asset_ins: Mapping[str, AssetIn],
     non_argument_deps: Optional[AbstractSet[AssetKey]],
-) -> Dict[str, In]:
+) -> Mapping[AssetKey, Tuple[str, In]]:
```

Comment:
It would be helpful to document the meanings of the values in this data structure. I'm having a little trouble following it.